keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -0,0 +1,521 @@
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The initial implementation of [dual-stack networking] in Kubernetes
+left the apiserver as a single-stack Service. This moves it to
+dual-stack (in dual-stack clusters).
+
+[dual-stack networking]: ../563-dual-stack/README.md
+
+## Motivation
+
+### Goals
+
+- Ensure kube-apiserver can listen dual-stack (subject to appropriate
+  feature gates):
+
+    - Allow specifying dual-stack `--bind-address` on the command
+      line.
+
+    - Autodetect dual-stack bind addresses if possible when no
+      explicit `--bind-address` is specified.
+
+- Allow kube-apiserver to advertise dual-stack functionality to other
+  components (subject to appropriate feature gates):
+
+    - Allow specifying dual-stack `--advertise-address` on the command
+      line (and likewise, autodetecting dual-stack advertise addresses
+      from the bind address when `--advertise-address` is not
+      explicitly specified).
+
+    - Publish dual-stack `EndpointSlice`s for the "`kubernetes`"
+      Service when dual-stack apiserver IPs are available.
+
+    - When configured with both dual-stack bind/advertise addresses
+      and dual-stack service CIDRs, create the "`kubernetes`" service as
+      `ipFamilyPolicy: RequireDualStack`, and assign it the first IP
+      address in both the IPv4 CIDR range and the IPv6 CIDR range.
+
+### Non-Goals
+
+- Trying to make `KUBERNETES_SERVICE_HOST` dual-stack.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - External Clients
+
+As a cluster administrator, I want my apiserver to be reachable by
+both IPv4-only and IPv6-only clients outside the cluster, without them
+needing to use a proxy or NAT mechanism.
+
+#### Story 2 - APIServer Tooling
+
+As a Kubernetes distributor or service provider, I want to be able to
+reliably determine the IP addresses being used by the apiserver in a
+user-created cluster, so that I can correctly create and configure
+related infrastructure (eg, TLS certificates, endpoints for external
+load balancers or DNS names, etc).
+
+#### Story 3 - Internal Clients (?)
+
+As a developer, I want to run a pod in a dual-stack cluster but only
+access the apiserver via the cluster's secondary IP family (eg, via
+IPv6 in an IPv4-primary cluster).
+
+(This has been brought up, but it is not clear that there is really a
```

Comment:
is not this one https://github.com/kubernetes/enhancements/issues/563#issuecomment-814560270?